pkg/controller/database/rdsinstance.go

```diff
@@ -0,0 +1,209 @@
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	errNotRDSInstance   = "managed resource is not an RDS instance custom resource"
+	errKubeUpdateFailed = "cannot update RDS instance custom resource"
+
+	errCreateRDSClient   = "cannot create RDS client"
+	errGetProvider       = "cannot get provider"
+	errGetProviderSecret = "cannot get provider secret"
+
+	errCreateFailed        = "cannot create RDS instance"
+	errModifyFailed        = "cannot modify RDS instance"
+	errAddTagsFailed       = "cannot add tags to RDS instance"
+	errDeleteFailed        = "cannot delete RDS instance"
+	errDescribeFailed      = "cannot describe RDS instance"
+	errPatchCreationFailed = "cannot create a patch object"
+	errUpToDateFailed      = "cannot check whether object is up-to-date"
+)
+
+// SetupRDSInstance adds a controller that reconciles RDSInstances.
+func SetupRDSInstance(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.RDSInstanceGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.RDSInstance{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.RDSInstanceGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), log: l.WithValues("connector", "rds")}),
+			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube client.Client
+	log  logging.Logger
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	c.log.Info("Connect", "ns/name", mg.GetNamespace()+"/"+mg.GetName())
+	cr, ok := mg.(*v1alpha1.RDSInstance)
+	if !ok {
+		return nil, errors.New(errNotRDSInstance)
+	}
+
+	p := &aliv1alpha1.Provider{}
+	err := c.kube.Get(ctx, meta.NamespacedNameOf(cr.Spec.ProviderReference), p)
+	if err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	rdsClient, err := rds.NewClient(ctx, string(s.Data["accessKeyId"]), string(s.Data["accessSecret"]), p.Spec.Region)
+	return &external{client: rdsClient, kube: c.kube, log: c.log}, errors.Wrap(err, errCreateRDSClient)
+}
+
+type external struct {
+	client rds.Client
+	kube   client.Client
+	log    logging.Logger
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	e.log.Info("Observe", "ns/name", mg.GetNamespace()+"/"+mg.GetName())
```

Comment:
Please use debug logs for this. Assume the controller will reconciling tens or hundreds of managed resources - I think a happy-path reconcile in which all goes to plan should result in zero info log lines. We can use Kubernetes events more generously.

Reply:
Also, managed resources are cluster scoped and will never have a namespace.

Reply:
Reopening this; I would really prefer us not to be producing an info log for every reconcile loop.

Reply:
This log has been removed.
Could you clarify your concern if not about this log?

Reply:
I see this specific has been removed - thank you. There are other info log lines at the beginning of the Connect, Create, Update, and Delete methods. Those `log.Info` lines should be either removed or switched to `log.Debug`.

The concern is that each individual reconcile loop should not be producing info logs during the happy path. This results in a lot of log noise in a busy Crossplane provider that reconciles many managed resources. It's also not actually that useful; most users of Crossplane aren't operators of Crossplane and thus can't see the logs of the Crossplane pods.

We prefer to emit Kubernetes events rather than logs - this way a Crossplane user can see what is going on with their RDS instance by typing `kubectl describe rdsinstance`. It is also okay to produce _debug_ level logs to help developers and operators debug Crossplane. We already emit both debug logs and Kubernetes events in `managed.Reconciler`, so I would argue they are not necessary here in the `managed.ExternalClient` implementation. [This document](https://github.com/crossplane/crossplane/blob/84b1c08d18900a99997a02e6417f5b63aeace3f3/docs/4_contributing/2_observability_developer_guide.md) contains more detail on our logging and observability approach.

Reply:
removed